File: src/leetcode/array/java/MinimumSizeSubarraySum.java
package leetcode.array.java;
/*
    like sliding window
    while loop,
    match, caculate result
    cut left element, left++
    keep going...

    notice that, not found should return 0
 */
public class MinimumSizeSubarraySum {
    public int minSubArrayLen(int s, int[] nums) {
        int res = Integer.MAX_VALUE;
        int sum = 0;
        int left = 0;

        for (int i = 0; i < nums.length; i++) {
            sum += nums[i];
            while (left <= i && sum >= s) {
                res = Math.min(res, i - left + 1);
                sum -= nums[left++];
            }
        }
        return res == Integer.MAX_VALUE ? 0 : res;

    }
}

File: src/leetcode/array/kotlin/MinimumSizeSubarraySum.kt
package leetcode.array.kotlin

/*
    Input: s = 7, nums = [2,3,1,2,4,3]
    Output: 2

    // sliding windows
    // for loop continue sum,
    // while left < i && sum >= s,
    // calate min distance
    // sum -= left
 */
fun minSubArrayLen(s: Int, nums: IntArray): Int {
    var sum = 0
    var left = 0
    var res = Int.MAX_VALUE
    nums.forEachIndexed { i, num ->
        sum += num
        while (sum >= s) {
            res = Math.min(res, i - left + 1)
            sum -= nums[left++]
        }
    }
    if (res == Int.MAX_VALUE) return 0 else return res
}